### DISCUSSION TITLE
Storing/Exporting local keypair

### DISCUSSION THREAD

**Discussion by 2-3-5-41** [libp2p/rust-libp2p#4376] [Q&A] [answered]
I want to know if there is a _recommended_ way to export/store a keypair in libp2p. I ask this as my current idea for a project requires a persistent identity for a user/peer even after they restart their client.

I've already looked at the `to_protobuf_encoding()` and `from_protobuf_encoding()` functions, however, `from_protobuf_encoding()` on my system seems to "corrupt" the `EdwardsPoint.Z`, assuming the Z field isn't supposed to look like this `FieldElement51([1, 0, 0, 0, 0])` after getting a keypair from a protobuf encoding.

**Comment by thomaseizinger** [comment, marked as answer]
> however, `from_protobuf_encoding()` on my system seems to "corrupt" the `EdwardsPoint.Z`, assuming the Z field isn't supposed to look like this `FieldElement51([1, 0, 0, 0, 0])` after getting a keypair from a protobuf encoding.

This sounds like a bug to me. Can you reproduce this with a unit test and open a PR?

> I want to know if there is a _recommended_ way to export/store a keypair in libp2p.

It is recommended to use the `{to,from}_protobuf_encoding` functions because those "tag" the secret key bytes with their type and allow us correctly load them again.

**Reply by 2-3-5-41** [reply]
How can I write a test to check if a decoded keypair is the same as the original? Does libp2p have an internal method of doing so, the only reason I found this was from printing out the original and decoded keypairs to my terminal to see if there were any differences.

**Reply by thomaseizinger** [reply]
The easiest way is to derive a public key and then a `PeerId` and compare those via `PartialEq`.

**Reply by 2-3-5-41** [reply]
I'm trying to make a unit test for this, but I'm getting this error;
```
error[E0432]: unresolved import `libp2p_identity::PeerId`
 --> identity/tests/protobuf_encode_decode_api.rs:1:32
  |
1 | use libp2p_identity::{Keypair, PeerId};
  |                                ^^^^^^ no `PeerId` in the root

error[E0599]: no function or associated item named `generate_ed25519` found for struct `Keypair` in the current scope
 --> identity/tests/protobuf_encode_decode_api.rs:5:28
  |
5 |     let keypair = Keypair::generate_ed25519();
  |                            ^^^^^^^^^^^^^^^^ function or associated item not found in `Keypair`
```

And my code looks like this;

```rs
use libp2p_identity::{Keypair, PeerId};

#[test]
fn to_from_protobuf() {
    let keypair = Keypair::generate_ed25519();
    let peer_id_before = PeerId::from(&keypair.public());

    let protobuf_encode = keypair.to_protobuf_encoding().unwrap();
    let keypair_from_protobuf = Keypair::from_protobuf_encoding(&protobuf_encode.as_slice()).unwrap();
    let peer_id_after = PeerId::from(&keypair_from_protobuf.public());

    assert_eq!(peer_id_before, peer_id_after)
}
```

I'm doing this in the `test` module folder in the identity crate/workspace member.

**Reply by 2-3-5-41** [reply]
I was able to recreate the unit test in a separate crate that uses `libp2p` as a dependency, and the test passes, however, I still wonder if that corrupted field affects the private key since we're only testing the `PeerId` that is derived from the public key.

Unless I have a gross misunderstanding of how the keypairs work.

**Reply by thomaseizinger** [reply]
The public key is derived from the private key so if the latter is corrupted, you should get a different public key and thus different `PeerId`.

The error above stems from a missing feature. You'd need to feature-gate your test on the `peerid` feature.

**Reply by 2-3-5-41** [reply]
I just realized there are already tests doing this in the `keypair.rs` file.

Anyway, this thread has answered my question, as well as set my mind straight on how keypairs work, thank you.
